RSpecのテストコードを書く練習をするPull Request

1mori opened the pull request:
## 関連ページ
https://leanpub.com/everydayrailsrspec-jp/read
https://classi.esa.io/posts/34238

## 最終更新日
2024/06/27

## やりたいこと
- RSpecを用いてメソッド単位でのテストコードを書く練習をする
  - WEbセキュリティ課題のテストコードはE2E形式なので、さらに細部に対してのテストを書く

## やったこと
- `User`に関するテストを5種類書いた
- テストを書くときは`described_class`を使用してテストコード内で`User`を書かないようにした
- AAAを意識したテストコードの書き方を実践してみた
https://tech.pepabo.com/2021/08/23/writing-unit-test-with-aaa/
- テスト文は日本語に変更した

## やらなかったこと
- `nickname`の存在確認に対するテスト
  - モデルでニックネームに対する制約は`uniqueness`であることのみが定義され、`presence`については書かれていない、なので存在確認のテストを書いたら弾かれたため

## 見てほしいところ
- 自分なりにAAAを意識して書いてみたので、改めてテストコードの読みやすさをみてほしい
- テストデータの挿入が長い文章が並んでいてわかりにくいなと感じるのですがこれはもう仕方ないですか？（Rubocop的にはOKみたいです）
```ruby
let(:correct_user) { described_class.new(email: 'test@example.com', password: 'password', password_confirmation: 'password', nickname: 'testuser') }
    let(:incorrect_email_user) { described_class.new(email: nil, password: 'password', password_confirmation: 'password', nickname: 'testuser') }
    let(:incorrect_password_user) { described_class.new(email: 'test@example.com', password: nil, password_confirmation: 'password', nickname: 'testuser') }
    let(:incorrect_confirm_password_user) do
      described_class.new(email: 'test@example.com', password: 'password', password_confirmation: 'wrong_password', nickname: 'testuser')
    end
```

## 仕様
- データベースはSQLiteを使用

1mori committed:
add: bundle install rspec_rails

1mori committed:
add: bin/rails generate rspec:install

1mori committed:
add: RSpecの出力をドキュメント形式に変更

1mori committed:
add: bin/rspecでRSpecテストを実行できるようにbinにファイルを追加

1mori committed:
add: 不要なSpecファイルが作成されないように設定を追加

1mori committed:
add: ユーザー登録に関するテストを追加

ユーザー情報において必須項目をまず羅列してみた
その後、一番基礎となるユーザー登録がうまくいくかのテストを書いた

1mori committed:
fix: 改行の修正

1mori committed:
change: subject, described_classを用いた書き方に変更

ユーザーを予め一人用意したかった
subject { described_class.new(...)} でサンプルユーザーを定義した

1mori commented on spec/models/user_spec.rb:
```diff
@@ -0,0 +1,31 @@
+require 'rails_helper'
+
+RSpec.describe User, type: :model do
+  subject do
+    described_class.new(
+      email: 'test@example.com',
+      password: 'password',
+      password_confirmation: 'password',
+      nickname: 'testuser'
+    )
+  end
```
あまり共通化はしないほうがいいとの教えを受けておいてなんですが、テストユーザーを一人定義する処理は簡潔な処理であり、かつ全てのUserのテストにおいて使用する見込みなので共通化しました。

kudoas replied:
[IMO] ここ結構議論があるんですけど、少なくともモデルを作るだけの目的で subject は使うことは少ないと思います。

https://blog.jnito.com/entry/2021/10/09/105651

あとこういったカスタマイズされた User オブジェクトを生成したい場合は FactoryBot を使うことが多いです。そうすると他のテストでも使い回すことが多くて便利です！今後出てくるので覚えておくと良いと思います。

https://github.com/thoughtbot/factory_bot/blob/main/GETTING_STARTED.md

kudoas replied:
[NITS] テストケースごとに subject で参照していますが、テストケースの中身を見る時に参照先をいちいち確認しないといけないので、名前つけたほうがいいです。

---

[MUST] 
そもそも正常系のオブジェクトを上書きして、validation をチェックするテストの書き方はあんまりよくないです。
それよりは、後述するように必要なオブジェクトをテストケースで生成する方がテストケースとして完結してわかりやすいと思います。

1mori reviewed (commented):
テストユーザーの定義についてコメントしました！

1mori committed:
add: Userモデルのに関するテストを追加

Eメールがない時、パスワードがない時、確認用パスワードと一致しないとき、ニックネームが重複する時に失敗するかどうかのテストを書いた

kudoas commented on .rspec:
```diff
@@ -1 +1,2 @@
 --require spec_helper
+--format documentation
```
[IMO] = in my opinion (自分はこう思うぐらいの指摘)

rspec のデフォルトで`--format documentation` をここで指定しなくていいと思います。
例えば CI 上で rspec を動かしたい時に、テストが通るか見たいだけなので、ドキュメントは表示しても意味ないからです。
ドキュメントとして手元で確認したいというモチベーションはGoodですが、デフォルトにするまでではないと思います。

1mori replied:
今回はEveryday Railsに書かれていた設定を手元の環境で実行する、という動作確認を含めてこのオプションの指定を行いました。

https://leanpub.com/everydayrailsrspec-jp/read#leanpub-auto-section-17:~:text=%E3%81%8C%E4%B8%80%E7%95%AA%E3%81%A7%E3%81%99%E3%80%82-,%E6%AC%A1%E3%81%AB%E3%80%81,-%E5%BF%85%E9%A0%88%E3%81%A7%E3%81%AF%E3%81%82%E3%82%8A%E3%81%BE
> 次に、必須ではありませんが、私はRSpecの出力をデフォルトの形式から読みやすいドキュメント形式に変更するのが好みです。これによってテストスイートの実行中にどのスペックがパスし、どのスペックが失敗したのかがわかりやすくなります。それだけでなく、スペックのアウトラインが美しく出力されます。予想していたかもしれませんが、この出力を仕様書のように使うこともできます。先ほど作成された .rspec ファイルを開き、以下のように変更してください。
> ```
> --require spec_helper
> --format documentation
> ```

書籍でも必須ではないと書かれていますし、私自身オプションの違いを出力で確認しましたが、別に見やすいと感じるほどじゃないなとは思っていたので、この行は消すことにします！

kudoas replied:
`--format documentaion` ではテストのドキュメントを仕様として確認することができるので、オプションとしては有用だと思います。何もオプションつけなければ落ちているテストしか表示されません。

> rspec のデフォルトで--format documentation をここで指定しなくていいと思います。

そういう意味では見たい時に使うのでOKで、デフォルトとして設定する必要はないかなと思ってのコメントでした。

1mori replied:
必要のないオプションではなく、使用したい時に適宜オプションをつけるのがいいということですね
今後必要だと感じた時に活用してみます！コメントありがとうございました🙇‍♂️

kudoas commented on bin/rspec:
[ASK] ＝ 質問
この binstub どうやって生成しました？生成するとどんなメリットがありますか？

1mori replied:
こちらは以下のコマンドで生成しました。
```
$ bundle binstubs rspec-core
```

こちらも上と同様、書籍に書かれていた内容です。

https://leanpub.com/everydayrailsrspec-jp/read#leanpub-auto-section-17:~:text=%E3%82%88%E3%81%86%E3%81%AB%E3%81%99%E3%82%8B-,Rails,-%E3%82%A2%E3%83%97%E3%83%AA%E3%82%B1%E3%83%BC%E3%82%B7%E3%83%A7%E3%83%B3%E3%81%AFBundler
> RailsアプリケーションはBundlerの使用が必須であるため、RSpecを実行する際は`bundle exec rspec`のように毎回 `bundle exec`を付ける必要があります。binstubを作成すると`bin/rspec`のように少しタイプ量を減らすことができます。binstubを作成する場合は以下のコマンドを実行します。

タイピング量が減らせるのがウリですが、正直一度入力すれば過去の入力を遡れるのでそこまで便利だとは感じなかったですね…

kudoas replied:
良いと思いました！

タイプ減らすぐらいなら自分も `bundle exec rspec ...` で直接打ってしまうことが多いですね。
過去のコマンド履歴も検索できるように peco で設定しています。

refs. https://github.com/peco/peco

1mori replied:
ありがとうございます！
peco はまだ使ったことがないので空き時間に調べてみます！

kudoas commented on spec/models/user_spec.rb:
```diff
@@ -0,0 +1,49 @@
+
+  # 重複したニックネームは無効な状態であること
+  it 'is invalid with a duplicate nickname' do
+    described_class.create!(
+      email: 'test2@example.com',
+      password: 'password',
+      password_confirmation: 'password',
+      nickname: 'testuser'
+    )
```
[ASK] ここで Userを作成すると users テーブルにレコードが保存されますが、そのデータはどうなると思いますか？
残ったままになるのか、誰かが削除してくれるのか？（それを担っているオプションがあるのか）調べてみてみてください。

1mori replied:
調べてきました！
（夕会で報告したのでメモとして残しておきます）

Rails_helper.rbの36行目の記述がテストコードで生成したデータを削除するかどうかを管理しています

https://gist.github.com/TiuTalk/551378d3451361be0212#file-rails_helper-rb-L36
```
config.use_transactional_fixtures = true
```

big-Miyabi commented on bin/rspec:
[ASK]
これは自動生成されたコードでしょうか？それとも自分で記述したものでしょうか？
引用元や参考にしたURLがあれば、コミットメッセージ内に記述が欲しいです🙏

1mori replied:
内容としては同様のものであると思うのでリンクの形で回答を繋げておきます！ https://github.com/1mori/haunted-blog/pull/2#discussion_r1656576238

書籍に書かれていたので実行したけど、別にいらないな〜というのが正直な感想です

big-Miyabi replied:
コミットメッセージにはなぜ必要なのかを書くと良いです！
「bin/rspecでRSpecテストを実行できるように」でもなんとなく意図は伝わりますが、「RSpecの実行を簡単にするため」のようにすると直接的で読み手に優しくなります

> コードには How
> テストコードには What
> コミットログには Why
> コードコメントには Why not
https://x.com/t_wada/status/904916106153828352

ただ不要と感じているとのことなので、消してしまうのが良いかと思いました

1mori replied:
了解しました！
今回は書籍に書いていたことをなんとなくでやってしまったところがあるので、次回以降なぜそれをするのか？を意識するようにします

big-Miyabi commented on config/application.rb:
```diff
@@ -20,5 +20,16 @@ class Application < Rails::Application
 
     config.time_zone = 'Tokyo'
+
+    Dir[Rails.root.join('lib', '**/')].each do |dir|
+      unless dir.match(/assets|tasks/)
+        config.autoload_paths << dir
+      end
+    end
```
assets や tasks を含んでいないlib以下のディレクトリを全て読み込んでいるように見えますが、ここはなぜ必要なのでしょうか？
見たところ、lib以下には`.keep`以外ファイルがなかったので、無くても良いのではと思いました

big-Miyabi replied:
また、コミットメッセージを読むと「不要なSpecファイルが作成されないように設定を追加」となっていますが、この部分は関係ないように見えます
コミットを分けて、意図がわかるようにコメントするなり、コミットメッセージで説明するなりして欲しいです

1mori replied:
このコミットを一つにした理由として、書籍内でRSpecに対する設定を行う箇所があったのですが、自環境の`config/application.rb`と比較したときに、不足している記述（今回であれば11行目以降）が全てRSpecに関する設定の記述だと思い込み、まとめて追加しコミットしたというのが理由です。
書き方が異なるのは、`config.autoload_lib`だとコードが通らなかったため、調べたときに出てきた`autolib_paths`の書き方に変更したためです。

設定についてよくわかっていない状態で追加したのもあり、今回の要件では必要ないということに気づいていませんでした。
該当行はこのあと削除しておきます。

https://leanpub.com/everydayrailsrspec-jp/read#leanpub-auto-section-17:~:text=%E3%81%AA%E3%82%8A%E3%81%BE%E3%81%97%E3%81%9F%E3%80%82-,%E3%82%B8%E3%82%A7%E3%83%8D%E3%83%AC%E3%83%BC%E3%82%BF,-%E3%81%95%E3%82%89%E3%81%AB%E3%80%81%E3%82%82%E3%81%86%E3%81%B2%E3%81%A8%E3%81%A4
> ジェネレータ
さらに、もうひとつ手順があります。 rails generate コマンドを使ってアプリケーションにコードを追加する際に、RSpec用のスペックファイルも一緒に作ってもらうようRailsを設定しましょう。
（中略）
config/application.rb を開き、次のコードをApplicationクラスの内部に追加してください。
```
 1 require_relative "boot"
 2 
 3 require "rails/all"
 4 
 5 # Railsが最初から書いているコメントは省略 ...
 6 Bundler.require(*Rails.groups)
 7 
 8 module Projects
 9   class Application < Rails::Application
10     config.load_defaults 7.1
11     config.autoload_lib(ignore: %w(assets tasks))
12 
13     config.generators do |g|
14       g.test_framework :rspec,
15         fixtures: false,
16         view_specs: false,
17         helper_specs: false,
18         routing_specs: false
19     end
20   end
21 end
```

kudoas commented on spec/models/user_spec.rb:
```diff
@@ -0,0 +1,49 @@
+  end
+
+  # Eメールがなければ無効な状態であること
+  it 'is invalid without an email address' do
+    subject.email = nil
+    expect(subject).to be_invalid
+  end
```
[IMO]

一回作ったオブジェクトを使い回してテストせずに、必要なオブジェクトだけそのテストで使うようにしたほうがいいです。

イメージこんな感じです。

```suggestion
  it 'is invalid without an email address' do
    user = User.new(email: nil, password: 'password', nickname: 'unique_nickname')
    expect(user.valid?).to be false
  end
```

1mori replied:
修正し、共通して以下のスタイルにしました。

```suggestion
  it 'is invalid without an email address' do
    user = described_class.new(...)
    expect(user.valid?).to be [true もしくは false]
  end
```

[ASK]

今回クラス名を書くときに`described_class`を使用したのですが、`User`と書いたほうがいいですか？
Rubocopでは`described_class`が使えるときに使わないと注意を出すらしいのですが、私の環境では出なかったので、どうすればいいか悩んでいます。

https://www.rubydoc.info/gems/rubocop-rspec/RuboCop/Cop/RSpec/DescribedClass

kudoas replied:
> 修正し、共通して以下のスタイルにしました。

修正したコードの参照か、コミットだと嬉しいです。
こう変えました！と言われても、実際にコードが変わってるかどうか判断できないからですね。
確認したところ、大丈夫でした！

> Rubocopではdescribed_classが使えるときに使わないと注意を出すらしいのですが、私の環境では出なかったので、どうすればいいか悩んでいます。
https://www.rubydoc.info/gems/rubocop-rspec/RuboCop/Cop/RSpec/DescribedClass

貼ってもらったリンクは rubocop-rspec という RuboCop の拡張 gem のオプションなので、設定しなければ出ないです。
https://github.com/1mori/haunted-blog/blob/main/.rubocop.yml#L2-L4

使いたい場合は設定すれば出ます。 https://github.com/rubocop/rubocop-rspec

ちなみに自分自身はこだわりないので、チームの方針やRuboCopのルールに合わせる or Userと書くと思います。

1mori replied:
> 修正したコードの参照か、コミットだと嬉しいです。
こう変えました！と言われても、実際にコードが変わってるかどうか判断できないからですね。

失礼しました🙇‍♂️　実はやり方がわかっていなかったので今から勉強してきます！

> 貼ってもらったリンクは rubocop-rspec という RuboCop の拡張 gem のオプションなので、設定しなければ出ないです。

なるほど！別のRubocopがあったのですね
この際使用感を確かめるために一度インストールしてみます！

> ちなみに自分自身はこだわりないので、チームの方針やRuboCopのルールに合わせる or Userと書くと思います。

どちらを使うかの意見もありがたいです、その時々で使い分けられるようにしておきます！

1mori replied:
前々回のコードの貼り方ですが、suggestionを使ってみました！
また一つやり方を覚えられました🙌

kudoas replied:
suggestion はレビュワーがコードを修正する箇所を指摘する時には使うんですけど、レビュイーで修正箇所を申告する時には使わないですね。

> 修正したコードの参照か、コミットだと嬉しいです。
こう変えました！と言われても、実際にコードが変わってるかどうか判断できないからですね。

suggestion を使った場合でも、コードを書き写してコメントに貼っていますよね。直接コードを参照しているわけではないので、指摘した内容を解消できてないです。
修正したコードの参照を貼る場合は permalink を使うと良いです。

https://github.com/1mori/haunted-blog/blob/ef8a787344bb2abf507b24e83b3e4215dc6eba5a/spec/models/user_spec.rb#L14-L20

github 上のコードを選択して、3点リーダーを押すとコピーできるので使ってみてください。

![image](https://github.com/1mori/haunted-blog/assets/45157831/979095bb-ac63-46ef-adbd-0cf5ff1e42c8)

1mori replied:
ありがとうございます！suggestionはレビュワー側が使う機能なのですね

最新コミットでCopy premallinkを用いて上に示していただいた例と同じ行を貼りました、この方法で次回から引用してみます！

https://github.com/1mori/haunted-blog/blob/38822ad57143aa34554309bf3d0d5259150f0114/spec/models/user_spec.rb#L14-L20

kudoas replied:
> suggestionはレビュワー側が使う機能

@1mori suggestion　= 提案ですからね〜！良いと思いました！

kudoas commented on spec/models/user_spec.rb:
```diff
@@ -0,0 +1,49 @@
+  end
+
+  # パスワードがなければ無効な状態であること
+  it 'is invalid without a password' do
```
これも同じ

1mori replied:
修正しました！
こちらのコメントと同様のスタイルです
https://github.com/1mori/haunted-blog/pull/2#discussion_r1658023747

kudoas commented on spec/models/user_spec.rb:
```diff
@@ -0,0 +1,49 @@
+
+  # 確認のパスワードが一致しなければ無効な状態であること
+  it 'is invalid with a mismatched password' do
+    subject.password_confirmation = 'pussword'
+    expect(subject).to be_invalid
```
これも同じ

1mori replied:
同じく修正しております！
https://github.com/1mori/haunted-blog/pull/2#discussion_r1658023747

kudoas commented on spec/models/user_spec.rb:
```diff
@@ -0,0 +1,49 @@
+    )
+  end
+
+  # Eメール、パスワード、ニックネームが有効な状態であること
```
it で説明書いてくれているので、コメントで日本語訳しなくていいですよ。
コメントと it の文章を同期させる運用が必要になるので、今後管理が大変になります。

1mori replied:
このあたり、書籍が日本語訳なのもあって書籍では上に日本語のコメント、下に`it 'example'`のような形で書かれていたのですが、後々日本語でいいよな…となって直すか悩んでいました。

翻訳された方自身が日本語でいいんじゃない？と書かれていること、
このあと書くテスト文を日本語に書きたいことを加味してテスト文を日本語に変更しておきます！

https://qiita.com/jnchito/items/42193d066bd61c740612#it-%E3%82%84-context-%E3%81%AB%E6%B8%A1%E3%81%99%E6%96%87%E5%AD%97%E5%88%97%E3%81%AF%E8%8B%B1%E8%AA%9E%E3%81%A7%E6%9B%B8%E3%81%8F%E6%97%A5%E6%9C%AC%E8%AA%9E%E3%81%A7%E6%9B%B8%E3%81%8F
> 日本人しか読まないテストコードなのであれば、日本語で書いても問題ないと僕は考えています。
しっくり合う英文を考えるのに時間を食うぐらいなら、その時間でテストコードを書いた方が時間の有効利用ができますからね。
それに、がんばって英文を書いても文法がめちゃくちゃで他の人が理解できない、なんていうリスクも出てくるかも・・・。

kudoas reviewed (commented):
最初なので少し範囲を広げて色々コメントしました

big-Miyabi reviewed (commented):
まだ最初の5コミットしか見ていませんが一旦コメントします〜

big-Miyabi commented on spec/models/user_spec.rb:
```diff
@@ -0,0 +1,49 @@
+    expect(subject).to be_valid
+  end
+
+  # Eメールがなければ無効な状態であること
+  it 'is invalid without an email address' do
+    subject.email = nil
+    expect(subject).to be_invalid
+  end
```
[IMO]
全体のテストの構造も意識できると良いです
色々書き方はありますが、コードのどの部分が Arrang / Act / Assert のどれに該当するのかをわかりやすくして、ファイル内で書き方を統一するとテストの可読性が上がります。

今回のケースはテスト自体が短いので既に読みやすくはありますが、練習としてAAAを意識して書いてみてください！

FYI: [AAA を意識して単体テストを書く - Pepabo Tech Portal](https://tech.pepabo.com/2021/08/23/writing-unit-test-with-aaa/)

1mori replied:
自分なりにAAAを意識した書き方に変えてみました！

- ユーザー登録をする場合と2人以上のユーザーがいる場合をcontextで分けました。
- それぞれのcontextの中で予め使うデータを定義し、テストでは実行文だけを書くように変更しました。

big-Miyabi replied:
以下でコメントしました！（こちらのスレッドはresolveします
https://github.com/1mori/haunted-blog/pull/2#discussion_r1658153029

1mori committed:
delete: --format documentationの設定を削除

書籍に書かれていたことの模倣として追加したが、特にこれが便利だ！とは思わなかったため

1mori committed:
delete: bin/rspeのファイルを削除

書籍上で`bin/rspec`が使えるようにと紹介されていたので使ってみたが、特段便利だとは感じなかったため

1mori committed:
change: テストコードの文章を日本語に変更

書籍の方法に倣って書いてみたが、日本人話者しかこのコード見ないので日本語でいいことに気づいたため

1mori committed:
change: 1つ1つののテスト内でテストユーザーを定義するようにした

1mori committed:
change: (user).to be_valid -> (user.valid?).to be true

修正後の書き方がわかりやすかったため

1mori committed:
change: 最初に続き、2~4番目のテストコードもスタイルを修正

1mori committed:
change: 重複したユーザーのテストコードのスタイルを修正

1mori committed:
change: AAAを意識した書き方に変更

最後のユーザーはテスト内に登場するユーザが2人でかつ、最初に正常なユーザーデータを作っておく必要があったため、
`before do`で正常ユーザーを作成してから重複したニックネームのユーザーの作成が失敗することを確かめた

1mori committed:
delete: 今回の要件で不要だったアプリケーションの設定を削除

削除した記述はEveryday Railsで書かれていた記述を参考にして書いた設定
https://leanpub.com/everydayrailsrspec-jp/read#leanpub-auto-section-17:~:text=%E3%81%AA%E3%82%8A%E3%81%BE%E3%81%97%E3%81%9F%E3%80%82-,%E3%82%B8%E3%82%A7%E3%83%8D%E3%83%AC%E3%83%BC%E3%82%BF,-%E3%81%95%E3%82%89%E3%81%AB%E3%80%81%E3%82%82%E3%81%86%E3%81%B2%E3%81%A8%E3%81%A4

内容としては、オートローダーによって読み込まれるファイルを制限するものだが、今回のアプリケーションの該当ディレクトリ(assets, tasks)には`.keep`ファイルしか置かれておらず、
制限の有無でRSpecの挙動に関わるものではないため

1mori committed:
change: 上4つのテストを「ユーザー登録をする場合」としてcontextでまとめた

big-Miyabi commented on spec/models/user_spec.rb:
```diff
@@ -0,0 +1,44 @@
+
+RSpec.describe User, type: :model do
+  context 'ユーザー登録をする場合' do
+    let(:correct_user) { described_class.new(email: 'test@example.com', password: 'password', password_confirmation: 'password', nickname: 'testuser') }
+    let(:incorrect_email_user) { described_class.new(email: nil, password: 'password', password_confirmation: 'password', nickname: 'testuser') }
+    let(:incorrect_password_user) { described_class.new(email: 'test@example.com', password: nil, password_confirmation: 'password', nickname: 'testuser') }
+    let(:incorrect_confirm_password_user) do
+      described_class.new(email: 'test@example.com', password: 'password', password_confirmation: 'wrong_password', nickname: 'testuser')
+    end
```
テストで使うものを上にまとめてしまうと目線が上下してしまい、また、どのテストでどれを使うのかがパッとわからなくなってしまうので、使うものだけでまとめるようにしてください！

例えば「Eメールがなければ無効であること」であれば、「Eメールがない場合」でcontextに括ってその中で使うものだけletで宣言すると良いと思います。

FYI: [willnet/rspec-style-guide: 可読性の高いテストコードを書くためのお作法集](https://github.com/willnet/rspec-style-guide?tab=readme-ov-file#describe%E3%81%A8context)

kudoas replied:
agree

1mori replied:
なるほど！さらに細かく分ければよかったのですね
一つ一つをcontextで括って書く書き方に変更します！

big-Miyabi commented on spec/models/user_spec.rb:
```diff
@@ -0,0 +1,44 @@
+
+RSpec.describe User, type: :model do
+  context 'ユーザー登録をする場合' do
+    let(:correct_user) { described_class.new(email: 'test@example.com', password: 'password', password_confirmation: 'password', nickname: 'testuser') }
```
> テストデータの挿入が長い文章が並んでいてわかりにくいなと感じるのですがこれはもう仕方ないですか？（Rubocop的にはOKみたいです）

do~endを使うと一行にしなくて済むようになります。それでも気になるようであれば適宜改行するのが良いかと思います
```rb
let(:correct_user) do
  described_class.new(email: 'test@example.com', password: 'password', password_confirmation: 'password', nickname: 'testuser')
end
```

ただ、ここら辺はFactoryBotを使うことで短く書けるようになります！

1mori replied:
ありがとうございます！
これを書いた後もしかしてFactoryBotなら簡潔に書けるかな？と休憩中に考えていたので、上記コメントでの指摘箇所と併せてFactoryBotで書き換えて再度pushします！

kudoas reviewed (commented):
コメントしました

1mori committed:
change: それぞれのテストをcontextで括る書き方に変更

今回の場合、複数のテストを括ると上でまとめてarrangeする形になってしまい、使用する変数がわかりにくくなってしまう
重複する定義もないため、一つのテストをcontextで囲うことで使用する変数がわかりやすくなった

1mori committed:
add: factory_bot_railsをインストール

1mori committed:
add: rubocop-rspecをインストール

1mori committed:
add: rubocop-rspecを使うように設定の有効化とテスト文制約の無効化

日本語でテスト文を書いた場合、`RSpec/ContextWording`で注意を出してくるが、今回は日本語で書くと決めているため、設定の無効化をした
参考URL: https://qiita.com/ikechan515/items/3bb1f9614be0ef65268e

1mori committed:
fix: タイプミス修正

1mori committed:
delete: factory-botの設定を無効にしていた行を削除

FactoryBotを使えるようにする

1mori committed:
add:  bin/rails g factory_bot:model user

Userモデルのテストで使うFactoryBotのファイルを生成した

1mori committed:
add: Factorybotでサンプルユーザーデータを作成

初期値として`correct_user`として定義していたデータを入れるようにした

1mori committed:
change: FactoryBotでユーザーデータの定義を書き換えた

1~4番目のテストはそれぞれがcontextで独立しているため、変数で定義を分ける必要がなくなった
なので、buildしたユーザーデータを格納する変数はuserに統一した
なぜならそちらの方が読みやすかったから

5番目のテストはユーザーデータが2つ登場する
検証するユーザーデータはnicknameが重複しているため、変数名はそのままにした

1mori committed:
fix: 最終行の改行追加

1mori commented on spec/models/user_spec.rb:
```diff
@@ -0,0 +1,49 @@
+    end
+  end
+
+  context '2人以上のユーザーが存在する場合' do
+    duplicated_nickname_user = FactoryBot.build(:user, email: 'test2@example.com')
+
+    before do
+      FactoryBot.create(:user)
+    end
+
+    it '重複したニックネームは無効であること' do
+      expect(duplicated_nickname_user.valid?).to be false
+    end
+  end
```
[ASK]

このテストではユーザーデータが2つ登場するため、変数名は`duplicated_nickname_user`のままにしたのですが、他と同様`user`でもいいですか？

big-Miyabi replied:
夕会でも話した内容をこちらにも

はい! userで良いと思います。スコープが広い場合は具体的な名前にしたほうが良いですが、今回は狭く、他に似た変数は存在しないので簡潔な名前でも良いと思います。

it内で使うものをletに、使わないものをbefore内に入れているのもgoodです👍

1mori reviewed (commented):
1点質問をしています！確認お願いします🙏

1mori committed:
add: FactoryBot.の記述を省略できる設定を追加

必ず追加してね！くらいの圧を感じた

参考URL: https://www.rubydoc.info/gems/factory_bot/file/GETTING_STARTED.md
> RSpec
> If you're using Rails, add the following configuration to `spec/support/factory_bot.rb` and be sure to require that file in `rails_helper.rb`:

1mori committed:
fix: buildで作成したユーザーデータをletで呼び出す書き方に変更

変更点は2つ
- 変数に格納していた処理をlet[:user] {...} という書き方に変更した
- FactoryBot.build などの書き方を前述の設定により省略して記載できるようにした

1mori committed:
add: frozen_string_literal: trueを追加

1mori committed:
add: ブログデータを追加

ブログの作成にはタイトル、本文、公開非公開、ユーザーの情報(user_id)が必要となる
そのため、users.rbで作成しているuserのサンプルデータをassociationで呼び出して、FactoryBot内の重複する記述を減らした

1mori committed:
add: ブログが正常に作成できるかどうかのテストを作成した

前回のコミットで作成したブログのサンプルデータを使用してブログの作成が正常に行われるかのテストコードを書いた

big-Miyabi commented on spec/models/user_spec.rb:
```diff
@@ -0,0 +1,49 @@
+  end
+
+  context '2人以上のユーザーが存在する場合' do
+    let(:user) { build(:user, email: 'test2@example.com') }
+
+    before do
+      create(:user)
+    end
```
beforeで1人目のユーザが作られた後にletの遅延評価により2人目のユーザ作られるので、実行順になるように記述を並び替えるとより良くなりそうです

また、do~end は {} に置き換え可能なので、letの記述に合わせて {} に変更しつつ一行で済ませると統一感が出て読みやすくなると思います

1mori replied:
なるほど、最初にletを登録しないといけないという固定観念に囚われていました
指摘箇所を変更し、pushしておきました！
https://github.com/1mori/haunted-blog/pull/2/commits/38822ad57143aa34554309bf3d0d5259150f0114

1mori committed:
change: beforeの書き方を変更し、実行順を前後させた

テストコードの実行としては
1. 1人目のユーザーをcreate
2. 2人目のユーザーをbuildする設定をletに登録する
3. テストが失敗することを確認する

となっているため、その順になるようにテストコードを書き換えた

big-Miyabi reviewed (approved):
LGTM!

kudoas reviewed (approved):
![LGTM](https://image.lgtmoon.dev/253758)
